pkg/storage/etcd3/store.go

```diff
@@ -0,0 +1,310 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"errors"
+	"path"
+	"reflect"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/conversion"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/storage"
+	"k8s.io/kubernetes/pkg/storage/etcd"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type store struct {
+	client     *clientv3.Client
+	codec      runtime.Codec
+	versioner  storage.Versioner
+	pathPrefix string
+}
+
+type elemForDecode struct {
+	data []byte
+	rev  uint64
+}
+
+func newStore(c *clientv3.Client, codec runtime.Codec, prefix string) *store {
+	return &store{
+		client:     c,
+		versioner:  etcd.APIObjectVersioner{},
+		codec:      codec,
+		pathPrefix: prefix,
+	}
+}
+
+func (s *store) Backends(ctx context.Context) []string {
+	resp, err := s.client.MemberList(ctx)
+	if err != nil {
+		glog.Errorf("Error obtaining etcd members list: %q", err)
+		return nil
+	}
+	var mlist []string
+	for _, member := range resp.Members {
+		mlist = append(mlist, member.ClientURLs...)
+	}
+	return mlist
+}
+
+func (s *store) Codec() runtime.Codec {
+	return s.codec
+}
+
+func (s *store) Versioner() storage.Versioner {
+	return s.versioner
+}
+
+func (s *store) Set(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
+	rev, err := extractAndResetRevision(s.versioner, obj)
+	if err != nil {
+		return err
+	}
+	data, err := runtime.Encode(s.codec, obj)
+	if err != nil {
+		return err
+	}
+	key = keyWithPrefix(s.pathPrefix, key)
+
+	// Note that if rev = 0, it means the key doesn't exist. In such case, it's equivalent to create.
+	txnResp, err := s.client.KV.Txn(ctx).If(
+		clientv3.Compare(clientv3.ModifiedRevision(key), "=", rev),
+	).Then(
+		clientv3.OpPut(key, string(data)),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	if !txnResp.Succeeded {
+		return storage.NewResourceVersionConflictsError(key, rev)
+	}
+
+	if out != nil {
+		putResp := txnResp.Responses[0].GetResponsePut()
+		return decode(s.codec, s.versioner, data, out, putResp.Header.Revision)
+	}
+
+	return nil
+}
+
+func (s *store) Get(ctx context.Context, key string, out runtime.Object, ignoreNotFound bool) error {
```

Comment:
seems not, please see:
```
For example, consider a client completing a write at time point 1 (t1). A client issuing a read at t2 (for t2 > t1) should receive a value at least as recent as the previous write, completed at t1. However, the read might actually complete only by t3, and the returned value, current at t2 when the read began, might be "stale" by t3.
```
t3>t2>t1
write-> value1-- t1
read -- begin at t2, complete at t3
when there has another write  completes(value2) in t2~t3, the read maybe get  value2. am i right?
@hongchaodeng